File: code/raytracer/python/scene.py
from utils import jsonfile2obj

from camera import *
from sphere import *
from lights import *


class Scene(object):
    def __init__(self, maxReflectionRecursions, camera, background, spheres, pointLights, ambientLights):
        self.maxReflectionRecursions = maxReflectionRecursions
        self.camera = camera
        self.background = background
        self.spheres = spheres
        self.objects = spheres
        self.pointLights = pointLights
        self.lights = pointLights
        self.ambientLights = ambientLights


def sceneHook(obj):
    if '__type__' in obj:
        if obj['__type__'] == "scene":
            return Scene(obj["maxReflectionRecursions"], obj["camera"], obj["background"], obj["spheres"],
                         obj["pointLights"], obj["ambientLights"])
        if obj['__type__'] == "camera":
            return Camera(obj["fov"], obj["position"], obj["up"], obj["target"], obj["near"])
        if obj['__type__'] == "sphere":
            return Sphere(obj["radius"], obj["position"], obj["materials"])
        if obj['__type__'] == "pointLight":
            return PointLight(obj["position"], obj["color"])
        if obj['__type__'] == "directionalLight":
            return DirectionalLight(obj["direction"], obj["color"])
        if obj['__type__'] == "ambientLight":
            return AmbientLight(obj["color"])
    return obj


def loadScene(sceneFile):
    return jsonfile2obj(sceneFile, object_hook=sceneHook)

File: code/raytracer/javascript/scene.js
function loadScene(scene)
{
    setCameraCoordinatesBasis(scene.camera);
    scene.getBackgroundColor = function(){
        if ("background_color" in this.params)
            return this.params["background_color"]
        return [0,0,0]
    };
    scene.getParam = function(paramName){
        if (paramName in this.params)
            return this.params[paramName]
        return None
    };
    scene.getAmbientLights = function(){
        var ambientLights = [];
        for(var i=0; i<this.lights.length; i++)
        {
            if (this.lights[i].__type__ == "ambient_light")
                ambientLights.push(this.lights[i]);
        }
        return ambientLights;
    };
    scene.getShadingLights = function(){
        var shadingLights = [];
        for(var i=0; i<this.lights.length; i++)
        {
            if (this.lights[i].__type__ != "ambient_light")
                shadingLights.push(this.lights[i]);
        }
        return shadingLights;
    }
    for(var i=0; i<scene.lights.length; i++)
    {
        if (scene.lights[i].__type__ == "directional_light")
            scene.lights[i].getDirection = function(p){ return this.direction;}
        else if (scene.lights[i].__type__ == "point_light")
            scene.lights[i].getDirection = function(p){ return normalize(sub(this.position, p));}
    }
    for(var i=0; i<scene.objects.length; i++)
    {
        if (scene.objects[i].__type__ == "sphere")
        {
            scene.objects[i].intersect = intersectSphere;
        }
    }
}